### DISCUSSION TITLE
How to pass variables to a page in a custom express route?

### DISCUSSION THREAD

**Discussion by theogravity** [vercel/next.js#20081] [Help]
I'm trying to implement the Ory Kratos auth service into next.js, and am having trouble translating the concepts from express to next.

In the example integration app located at:

https://github.com/ory/kratos-selfservice-ui-node

The registration route looks like this:

`app.get('/auth/registration', registrationHandler)`

The project has a registration handlebars template, which I would assume in next.js goes under `/pages/auth/registration.tsx` (after translating the handlebars stuff to react).

Where I'm stuck on is implementing the `registrationHandler` logic.

https://github.com/ory/kratos-selfservice-ui-node/blob/master/src/routes/registration.ts

In my next.js sample app on the server-side, I have:

```js
// this is a simplification of the code
import next from 'next'
import { parse } from 'url'
import express from 'express'

const nextApp = next({ // .... })

const handle = nextApp.getRequestHandler()

await nextApp.prepare()
const server = express()

const kratos = new PublicApi(new Configuration({ basePath: serverConfig.getAssertValue('base_url') }));

server.get('/auth/registration', (req: IRequest, res: Response, next: NextFunction) => {
   // code from registration.ts
})

server.all('*', (req: Request, res: Response) => {
  const parsedUrl = parse(req.url!, true)
  return handle(req, res, parsedUrl)
})
```

The problem I'm having is `registration.ts` has a `res.render` for rendering the handlebar template, and passes in a bunch of data to it.

```js
      // Render the data using a view (e.g. Jade Template):
      res.render('registration', {
        ...flow,
        oidc: methodConfig(flow, 'oidc'),
        password: methodConfig(flow, 'password')
      });
```

 How do I get the same behavior in next where I can pass variables to the page and have it be rendered by next? 

https://nextjs.org/docs/basic-features/data-fetching

Seems to indicate that the logic goes directly into the `/pages/auth/registration.tsx` file, but the docs seem to talk about it in the context of an API fetch, and not in a local pass from server to page situation.

**Comment by theogravity** [comment]
This looks like the answer, not sure if there are better solutions:

You set data in `res.locals`, and access it on the client via `Page.getInitialProps()`.

https://stackoverflow.com/questions/63642108/how-to-pass-something-like-visitor-hits-from-server-express-js-to-next-js
https://stackoverflow.com/questions/62736013/pass-value-from-server-side-to-client-side-in-next-js-using-getinitialprops

**Reply by meotimdihia** [reply]
Do you implement Next.js + Ory Kratos successfully?

**Reply by theogravity** [reply]
I got sidetracked with something else that I couldn't remember the end result.

**Reply by meotimdihia** [reply]
Thanks for the answer.
